woocommerce/woocommerce-ios
# Issue/5159 add product category filter

Part of https://github.com/woocommerce/woocommerce-ios/issues/5159

## Description
In the context of https://github.com/woocommerce/woocommerce-ios/issues/5159, we implement the UI for the  product filter list view, and the logic behind. We add a new row for Product Category filtering in the `FilterListViewController`, so the user can access the product categories list and select a category so the products list can be filter by category as well. However, please note that the `ProductCategoryFilterListViewController` that should hold the product categories list is not added yet, to keep the PR size in an acceptable level.
Furthermore, in this PR the product category filter logic is added to the relevant elements (filter view models, product list and storage) and the `ProductCategory` type is adapted to play as a filter. Existing unit tests are enhanced and new ones added, covering the new features added (feature flag, product category filter)
Lastly, a feature flag is added so the feature can be hidden in Release builds, as the feature is not yet finished.

## Changes
- Add feature flag for this case
- Make `ProductCategory` conform to `FilterType`
- Add a view model for Product Category in FilterListViewController so the row can be shown if the feature flag is enabled
- Handle clear filter logic also for Product Category
- Add a case for categories in `FilterListValueSelectorConfig`, so it can react to the selection (for now a no-op until the UI for the product categories is added)
- Add string to localizable strings
- Add `ProductCategory` parameter to `AppSettingStore`

## Testing
With feature flag enabled:
- Open the app
- Go to Products tab
- Open Filter
- Product Category row should be shown at the end, without any specific category selected (Any):

<img src="https://user-images.githubusercontent.com/1864060/138415960-8b36e51e-af52-467b-9bad-4e3b6db889a4.png" width="320" height="693">

Please check that the existing filter flow works as expected, and the new addition did not break anything:
- Adding a filter option
- Clearing filter options
- Showing products with filters

With the feature flag disabled:
- Disable the filter by categories feature by doing it in `DefaultFeatureFlagService` on by running the app in a release version
- Open the app
- Go to Products tab
- Open Filter
- Product Category row should **not** be shown

<img src="https://user-images.githubusercontent.com/1864060/138417076-9ee4e7bf-7706-4e10-9d44-bd18cb161756.png" width="320" height="693">

As in the previous case, please check that the existing filter flow works as expected, and the new addition did not break anything:
- Adding a filter option
- Clearing filter options
- Showing products with filters

## Next Steps
- Show filter product category list view
- Handle selection: UI, storage
- Filter product list by category: networking


Update release notes:
